packages/pigeon/lib/pigeon_lib.dart

```diff
@@ -618,6 +637,7 @@ class _RootBuilder extends dart_ast_visitor.RecursiveAstVisitor<Object?> {
     final dart_ast.FormalParameterList parameters = node.parameters!;
     late String argType;
     bool isNullable = false;
```

Comment:
Why is the your new helper class "TypeArgument" rather than a more general "TypeSpecification" (or whatever) so that it can be used for the main type itself? It seems odd to have exactly the construction you need to make these three variables also a single variable, which they conceptually are, but not use it because the class is described as being for the arguments rather than for types in general.

Reply:
I do introduce the concept of the `TypedEntity` in a later PR which allows me to operate on Fields and TypeArguments.  I've been just implementing this as I've needed it.  Once everything is done with the 1.0 features it might be worth revisiting this since there is still some ill fitting stuff like using Fields for return values.  Maybe it means shifting to a more generic concept, or maybe introducing a ReturnType class that implements `TypedEntity`.

Reply:
> I've been just implementing this as I've needed it.

But you already do need it in this PR, thus my comment: you have several places where you have exactly the set of things that the new type is meant to encapsulate, and you're now fixing the fact that that they aren't grouped, but only for an arbitrary subset instead of all cases. And then mixing the new (better) way and the old way in the same pieces of code.

Reply:
I just looked, and `TypedEntity` doesn't actually address my concern here. It doesn't fix the fact that `Field` is storing its type as three separate parameters, it just slaps an interface over it. I'm not suggesting that `Field` should be-a `TypeSpecification`, I'm saying it should have-a `TypeSpecification` instead of three separate members.

Reply:
I don't actually need it.  I've implemented the feature without it and without duplicating any code.  The reason I factored out the TypedEntity is because I eventually needed to it create logic that would operate over them both so I didn't have to duplicate it.  At this point it's more of a style issue imo.  I think refactoring to a ReturnType would make a lot of sense.  That way I could remove the last of the references to 'void' (I was able to remove those for arguments in a later PR).

Reply:
> I don't actually need it. I've implemented the feature without it

Obviously I don't mean the code doesn't work without it. I hope we can agree that the purpose of a code review is not to just make the code run though.

> and without duplicating any code.

This is objectively not the case:
`TypeArgument:`
```
class TypeArgument {
  [...]
  /// A string representation of the base datatype.
  final String dataType;

  /// The type arguments to this [TypeArgument].
  final List<TypeArgument>? typeArguments;

  /// True if the type is nullable.
  final bool isNullable;
```

`Field`:
```
class Field extends Node {
  [...]
  /// The data-type of the field (ex 'String' or 'int').
  [...]
  /// True if the datatype is nullable (ex `int?`).
  bool isNullable;
  [...]
  List<TypeArgument>? typeArguments;
```

The code this comment is attached to:
```
    late String argType;
    bool isNullable = false;
    List<TypeArgument>? argTypeArguments;
```

You have at least three different places just in the context of this PR alone where you have a collection of the three elements of the type expressed as three separate variables. One of them is declaring exactly the type that groups them, but artificially restricting its use. That creates both unnecessary duplication, and inconsistency, both of which hurt readability and maintenance.

> At this point it's more of a style issue imo.

I'm curious if you would argue that the difference between these two (slightly pseudo-code) implementations of a path are just a style issue:

```
class Point {
   int x;
   int y;
}

class Path {
  List<Point> points;
}
```

vs.

```
class SubsequentPoint {
  int x;
  int y;
}

class Path {
  int startX;
  int startY;
  List<SubsequentPoint> otherPoints;
}
```

Reply:
I think we look at things a bit differently.  Code is either defining how data resides in memory or is operating on that data.   Not duplicating yourself is only important for the actual operating logic since that's where maintenance and bugs come from.  The structure definitions are just a means to create hierarchy and organization to help people understand the data and drive the goal of not repeating your logic.  The structures obviously shouldn't try to confuse the data either though.

So you'd like this?
```dart
class TypeDeclaration {
  bool isNullable;
  String name;
  List<TypeDeclaration> typeArguments;
}

class Field {
  TypeDeclaration typeDeclaration;
}
```

I think that's fine.  I can make that change.  In terms of my goal of not duplicating logic, the follow is is just as good.  I do think it's just style.  Neither stops us from sharing logic and neither is confusing.  I would say your suggestion is slightly more clear since it involves less concepts.

```dart
interface TypedEntity {
  bool isNullable;
  String name;
  List<TypeArgument> typeArguments;
}

class TypeArgument implements TypedEntity {}

class Field implements TypedEntity {}
```

Reply:
(Although, one benefit of the later is that the concepts more closely match the elements in the language, so more of a wash on clarity and we still might need a TypedEntity that just returns a TypeDeclaration)

Reply:
Okay, I did the refactor, but I realized that it's going to be a headache to rebase the depending PRs.  Mind if I revert this and punt on this until the other things land?  I had to use the refactor tools to manage this safely.

Reply:
> Not duplicating yourself is only important for the actual operating logic since that's where maintenance and bugs come from. The structure definitions are just a means to create hierarchy and organization to help people understand the data and drive the goal of not repeating your logic.

Bugs come from people, and maintenance is done by people. Making code easier for people to understand makes maintenance easier, and bugs less likely.

> So you'd like this?

Yes. Types are a core object in Pigeon's logic, and having the components of a type clearly, distinctly encapsulated in an object that define a type makes it easier to understand logic that is operating on types.

> I do think it's just style. Neither stops us from sharing logic and neither is confusing.

I find it confusing that a core piece of data for your domain doesn't have a clear object, and that as a result I have to track a triad of variables as being related in multiple sections of code (just as, to go back to my simple example, I would find it confusing and problematic if a codebase in which a point was a core concept had no Point class, and just always used separate x and y variables).

And I find it very confusing that you are expressing the concept of "a thing that has a type" not via a has-a of a Type, which seems like the very obvious mapping, but via interface inheritance of a basket of data.

>  I would say your suggestion is slightly more clear since it involves less concepts.

If even you--as the person most familiar with the code as is because you wrote it all, and thus essentially guaranteed to find your version more readable than anyone else will find it--find my version slightly clearer, and I find my version much clearer, isn't that pretty good evidence that it is, in fact, more readable/understandable?

(There are also other effects. E.g.:
- Your version requires `toString`, conversion to a Dart string, and really any other logic that operates specifically on a type to be either duplicated in each implementation (going against your argument that your version is just as good when it comes to avoiding duplication), or made a freestanding function that takes a `TypedEntity` argument even if it would be clearer as a class method.

  In fact, looking at the next PR again: look closely at [this](https://github.com/flutter/packages/pull/418/files#diff-bda4e51dd560ea4eb7a3ea3dbca764dfb14d3014da1df8955bda4ff864c3603cR112) and [this](https://github.com/flutter/packages/pull/418/files#diff-bda4e51dd560ea4eb7a3ea3dbca764dfb14d3014da1df8955bda4ff864c3603cR147). You have already had to duplicate logic, **and even in that one single line of duplicate logic you already have a bug**.
- Your version does not allow extracting the type of another object like a Field. In fact, you list:
  > we still might need a TypedEntity that just returns a TypeDeclaration
  
  as a downside of my approach, but that's not a concept that's even expressible in your version. You have to always pass around the entire object (by interface) just to operate on the type. And that's not purely stylistic; you couldn't, for instance, have a set of types, or a collection keyed by types.)

> Okay, I did the refactor, but I realized that it's going to be a headache to rebase the depending PRs. Mind if I revert this and punt on this until the other things land?

But the reason it's hard to rebase the other PRs is because you're adding even more code that uses this and more layers around it (like `TypedEntity`). All of the problems I have with the type code in this PR are going to be further compounded in the subsequent PRs, and I wouldn't be able to suggest related improvements in any of that other new code either.

Reply:
Sorry, just seeing this now.

Well you raise some interesting points.  It seems to be at odds with the age old advice of the "Law of Demeter" which says we should prefer 'field.isNullable` over `field.type.isNullable`.

How about this as a compromise:
```dart
interface TypedEntity {
  bool isNullable;
  String name;
  List<TypeDeclaration> typeArguments;
}

class TypeDeclaration implements TypedEntity {
  bool isNullable;
  String name;
  List<TypeDeclaration> typeArguments;
}

class Field implements TypedEntity {
  TypeDeclaration type;
  bool get isNullable => type.isNullable;
  String get name => type.name;
  List<TypeDeclaration> get typeArguments => type.typeArguments;
}
```

This gives you the structure you want, makes Demeter happy and keeps the interface stable across the backlog of PRs we have.  We can decide if we want to eliminate TypedEntity in the future.

Just to verify, this is the last issue to resolve on this PR, right?  If you don't like this idea let's try to have a chat tomorrow.  I've started working on generating code for multiple arguments since this is at loggerheads.  The sooner we resolve this the less work I 'll have reconciling the existing code.

Reply:
> Well you raise some interesting points. It seems to be at odds with the age old advice of the "Law of Demeter" which says we should prefer 'field.isNullable`over`field.type.isNullable`.

I've not personally seen someone argue before that the LoD should be so strictly applied that simple data objects should never be used directly, but... 

> How about this as a compromise:

Sure, if you feel strongly about it, wrapping all of the accesses is fine. However you should be using different names for `TypedEntity`'s `name` getter wrapper since it's in the namespace of the enclosing class, not the type object, and is ambiguous in that context. (In particular, you're going to immediately have a bad time in `Field`.)

> Just to verify, this is the last issue to resolve on this PR, right?

As far as I know, yes.

Reply:
Okay, done.  PTAL.